### PR TITLE
ZEN 544: Step Definition Library Updates

### PR DESCRIPTION
**Pull request by mpcarolin**
**Ticket**: [ZEN-544](https://jira.simpleviewtools.com/browse/ZEN-544)

## Context/Goal
* implement a library of step definitions that can be reused in any feature for any project
* implement child/ancestor behavior
* implement dynamic "$world" variable in feature steps

## Updates
* `repos/testUtils/parkin/instance.js`
  * updated to ensure the parkin instance has the world object
* `repos/testUtils/playwright/**.js`
  * implemented some helpers wrapping playwright code
* `repos/testUtils/steps/given/isAncestor.js`
  * step definition that assign an ancestor, or throws if that ancestor doesn't exist
* `repos/testUtils/steps/given/isAncestorTo.js`
  * similar to `isAncestor`, but checks that a descendant exists as well
* `repos/testUtils/steps/given/openApp.js`
  * opens the app as defined in the herkin config
* `repos/testUtils/steps/given/openUrl.js`
  * opens the url, and checks for dynamic variables
* `repos/testUtils/steps/then/assertElementCount.js`
  * checks that an element exists a certain number of times on the DOM
* `repos/testUtils/steps/then/containsText.js`
  * added docs , converted to cucumber expression
* `repos/testUtils/steps/then/descendentContainsText.js`
  * checks that a given descendant contains the text
* `repos/testUtils/steps/then/inputContainsText.js` 
* `repos/testUtils/steps/then/isAttached.js`, `isDetached.js`, `isVisible.js`, `isHidden.js`
  * checks that the element is in the given state
* `repos/testUtils/steps/then/pageTitleIs.js` and `pageTitleIsNot.js`
  * updated the docs, updated with cucumberExpressions
* `repos/testUtils/steps/when/checkDescendent.js`
  * checks a descendent checkbox
* `repos/testUtils/steps/when/checkElement.js`
  * checks a checkbox
* `repos/testUtils/steps/when/clickDescendent.js`
  * clicks the descendent
* `repos/testUtils/steps/when/clickElement.js`
  * added docs and cucumber expressions
* `repos/testUtils/steps/when/pressKey.js`
  * presses the key
* `repos/testUtils/steps/when/setInputText.js`
  * sets input with text
* `repos/testUtils/steps/when/wait.js`
  * waits a specified number of seconds
* `repos/testUtils/steps/when/waitForNavigation.js`
  * waits for the page to load
* `repos/testUtils/steps/when/waitForState.js`
  * waits for the element to be in a state
* `repos/testUtils/support/getClientWorld.js`
  * helper for acquiring the consumer's world object
* `repos/testUtils/support/validate.js`
  * validates the world object
* `repos/testUtils/support/world.js`
  * the herkin world, merged with the client world
* `tasks/definitions/cucumber/run.js`
  * added flags for: tag filtering, bailing, slow mo, debug, and I shortened the timeout since it actually doesn't need to be super long 
* `tasks/utils/envs/replaceTemplateVars.js`
  * helper replacing the `herkin.config.js` `app.url` template variable with values like the project git branch
* `tasks/utils/envs/setMountEnvs.js`
  * extracted this helper to its ow nfile and updated to work with the app.url env
* `tasks/utils/playwright/metadata.js`
  * fixed bug where it wouldn't throw an error if the `browser-meta.json` file did not exist already on your system. Now, it will not throw, and it will just safely generate it.

### Setup
* Ensure your keg hub branch is on develop and has the latest changes
* `keg herkin && keg pr 23`
* `keg evf && keg pr 122 && keg evf start`
* `keg herkin start --from keg-herkin:zen-544-image-branch`

### Running tests
* `keg herkin cr test`
* Verify that all the tests for the tap pass

> these tests are defined and located in the tap's branch here: https://github.com/simpleviewinc/tap-events-force/pull/122

### CODE REVIEW

File: repos/testUtils/support/getClientWorld.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+const { tryRequireSync } = require('@keg-hub/jsutils/src/node')
+const path = require('path')
+
+/**
+ * Searches the client's support directory for a world export
+ * @return {Object?} - the client's world object, or undefined if it does not exist
+ */
+const getClientWorld = () => {
```
Comment (lancetipton) [inline repos/testUtils/support/getClientWorld.js:8]: This is a super clean implementation. Not an easy thing to do, but you solved it well

File: repos/testUtils/support/world.js [hunk 2]
```diff
@@ -1,9 +1,18 @@
...
+ */
+const HerkinWorld = deepMerge({
+  app: {
+    url: process.env.HERKIN_APP_URL
+  }
+}, getClientWorld())
 
-setWorldConstructor(HerkinWorld)
\ No newline at end of file
+setWorldConstructor(function (options={}) {
```
Comment (lancetipton) [inline repos/testUtils/support/world.js:14]: What is this doing exactly? Do we still need it.
It's importing form `cucumber`, which doesn't make sense to me?

File: repos/testUtils/steps/when/waitForNavigation.js [hunk 3]
```diff
@@ -1,8 +1,13 @@
 const { When } = require('HerkinParkin')
-const { getBrowserContext } = require('../../support/setup')
+const { getBrowserContext } = require('HerkinSetup')
 const { getPage } = getBrowserContext()
 
-When(/I wait for the page to load/, async () => {
+const waitForNavigation = async () => {
```
Comment (lancetipton) [inline repos/testUtils/steps/when/waitForNavigation.js:5]: Missing JSDoc Notes

File: repos/testUtils/steps/when/clickDescendent.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+const { When } = require('HerkinParkin')
+const { getElement } = require('HerkinPlaywright')
+const { checkForAncestor } = require('HerkinSupport/validate')
+
+const clickDescendent = async (selector, world) => {
```
Comment (lancetipton) [inline repos/testUtils/steps/when/clickDescendent.js:5]: Missing JSDoc Notes

**Review by lancetipton — CHANGES_REQUESTED** [review]
This is a really solid PR.
There's a lot here, but my first review of the code almost everything looks good.
There's a few minor things to look at / discuss?

I'm still testing things out, but I wanted to get this back, so you could look at what I mentioned.

thanks

File: repos/testUtils/support/world.js (hunk 2 above)
Comment (mpcarolin) [reply repos/testUtils/support/world.js:14]: not needed anymore, removed

File: repos/testUtils/steps/when/waitForNavigation.js (hunk 3 above)
Comment (mpcarolin) [reply repos/testUtils/steps/when/waitForNavigation.js:9]: updated

File: repos/testUtils/steps/when/clickDescendent.js (hunk 4 above)
Comment (mpcarolin) [reply repos/testUtils/steps/when/clickDescendent.js:10]: updated

**Review by lancetipton — APPROVED** [review]
This is absolutely awesome. Great work man. Thank you for doing this.

**Pull request merged**